states.file: Adding a keyvalue function

teizz opened the pull request:
### What does this PR do?

This PR adds a new function to the file state-module, including documentation of all its parameters. It also includes the addition of an integration test to test the new behaviour added. 

### What issues does this PR fix or reference?

No open issues specifically. 

### New Behavior

Using the new keyvalue_list function (or its keyvalue shorthand version), it's possible to make and maintain changes to files. Some of the features of this function are:

- uncommenting default lines, so all settings don't show up at the top/bottom of a file.
- removing superfluous lines which might have otherwise overridden a setting.
- matching either (or both) the key or value part of a setting case-insensitively as to not make unnecessary changes.
- append or prepend a key/value setting if it wasn't found in the file.
- fully supports a test=True run for detecting potential changes before actually committing them.

### Tests written?

Yes. There is an additional integration test which checks if a commented out line is commented in and has its value updated to reflect the requested setting.

### Commits signed with GPG?

No

### Rationale

Although it is already possible to (un)comment lines, append text, and replace lines through other functions in the file state-module, in some cases it would result in an sls file that's hard to read or a not-quite satisfactory result.

An example. Lots of products come with default configuration files which have the defaults commented out. Not all defaults might be in there though.

- `file.replace` used to be my go-to function, it is capable appending or prepending if the line is not found, can do regex, accepts regex-flags to do case-insensitive matching etc. However, unless one want to go for negated forward lookups, a line would always match, always get replaced and thus always trigger a change.
- `file.append`, and `file.blockreplace` can't really insert at an arbitrary location.
- `file.line` which got added in Beryllium won't replace a line which matches but already has the replacement content, and handles regex. But it can't easily match case-insensitively, and needs a location in order to 'ensure' a line exists.

I've been leaving a trail behind at different companies implementing the functionality in this PR in custom '_states' directories. All in different states of bug-fixing by now I'm sure.
During Saltconf18 talk of this came up and it was suggested I'd try to submit it 'upstream', since it would also be helpful in some of the Salt for Secops check/remediation states.

Commit:
Adding a keyvalue method to the file state module

For making changes to files based on a key/value pair.
This commit also includes addition of an integration test to test
the new behaviour added.

Commit:
fixing as many PyLint errors as possible

Commit:
mistook a blankline lint error for lack of blanklines between functions

Commit:
Lint error suggested salt.utils.fopen()

It should probably have suggested salt.utils.files.fopen() since
using salt.utils.fopen() breaks when running it through the
intergration test. It did seem to work when running it as a normal
state module though.

This should fix the additional lint errors.

Commit:
trying to fix a lint error somewhere else in the file I touched

teizz commented:
I've gone ahead an put 'versionadded' as 'Neon' for a placeholder in the docs, I suppose that'd need updating should this PR get merged.

Commit:
Trying to fix the windows2016-py2 build

Something on the windows2016-py2 build is throwing in what seems
like utf-16le encoding shenanigans in somewhere, but only for the
line actually being replaced, not any of the ones already there.
This obviously (and correctly I think) fails the test.

PermitRootLogin
becomes
P\x00e\x00r\x00m\x00i\x00t\x00R\x00o\x00o\x00t\x00L\x00o\x00g\x00i\x00n\x00

Also trying to be more OS friendly by using their native line
separators.

teizz commented:
The windows builds that are failing seem to err on 3 tests unrelated to this PR.
- integration.states.test_file.FileTest.test_file_managed_onchanges
- integration.states.test_file.FileTest.test_file_managed_prereq
- integration.states.test_file.FileTest.test_test_managed

The windows2016-py3 seems to also err on an additional file.managed test:
- integration.states.test_file.FileTest.test_managed_unicode_jinja_with_tojson_filter

I suspect they might have been failing at the last merge touching the file state module as well. At least merge of PR https://github.com/saltstack/salt/pull/49832 seems relevant to running in test, but I can't find even the last Jenkins build-log of when either the file state or test was touched.

cachedout commented on salt/states/file.py:
```diff
@@ -4570,6 +4570,372 @@ def replace(name,
     return ret
 
 
+def keyvalue_list(
```
The name of this function feels a little strange to me. Would it be more straightforward if it were just `file.keyvalue`?

teizz replied:
It is, and that's exactly what I'm doing in the other function below called exactly that. (https://github.com/saltstack/salt/pull/50627/files#diff-3b447dc835943beb1326f98f684fe783R4905-R4936)

 I'd written the first later, then figured it'd be nice to give a list. Since I didn't want to call the original function over and over I'd figured the shorthand version could instead call the list version.

I'd be happy calling it ``file.keyvalues``, and do away with the shorthand version though.

cachedout replied:
That makes a lot of sense. Excellent idea!

Commit:
Merge branch 'develop' of https://github.com/saltstack/salt into develop

Commit:
Combined file.keyvalue_list and file.keyvalue

integration.states.test_file.FileTest.test_keyvalue was already
testing the file.keyvalue function since it used file.keyvalue_list
already.

Now there is only one function; file.keyvalue which takes either
2 parameters (key, and value) or a dictionary key_values to handle
multiple key values in a single file all at once.

Commit:
Fix linter and docstring errors

Commit:
Merge branch 'develop' of https://github.com/saltstack/salt into develop

Commit:
fixing another linter warning

teizz reviewed (commented):
I think the additional commits resolve https://github.com/saltstack/salt/pull/50627#discussion_r236339886.

Commit:
Merge branch 'develop' of https://github.com/saltstack/salt into develop

Commit:
Merge branch 'develop' into develop

teizz committed:
Merge branch 'develop' into develop

cachedout commented on salt/states/file.py:
```diff
@@ -4570,6 +4570,375 @@ def replace(name,
+
+    search_only : False
+        Only search, do not actually replace. This is effectively a dry-run.
+        Functionally almost the same as running the state with test=True, which
```
Does it make sense to have a separate flag then? What exactly distinguishes this from `test=True`?

teizz replied:
You are absolutely right @cachedout ! Admittedly, I originally wrote this function as an execution module where the need for a ``search_only`` parameter made sense just like it does in the ``file.replace`` execution module.

Back then it was a nice way of seeing which minions would end up with changes and which had already had their config files updated. Ultimately it seemed to make a lot more sense to put this into a state module. I suppose I documented it's behaviour and never realised it really should be taken out.

I'll go ahead and clean this up.

cachedout replied:
Great! Thanks!

teizz committed:
Merge branch 'develop' of https://github.com/saltstack/salt into develop

teizz committed:
Cleanup search_only parameter

Left over from when this function was still part of an execution
module. It got converted when it made a lot more sense to have it
in an actual state module

s0undt3ch commented:
Port to master https://github.com/saltstack/salt/pull/56770
